Zend/zend.c

```diff
@@ -1892,3 +1860,29 @@ ZEND_API void zend_map_ptr_extend(size_t last)
 		CG(map_ptr_last) = last;
 	}
 }
+
+ZEND_API void zend_alloc_ce_cache(zend_string *type_name)
+{
+	if (ZSTR_HAS_CE_CACHE(type_name) || !ZSTR_IS_INTERNED(type_name)) {
+		return;
+	}
+
+	if ((GC_FLAGS(type_name) & IS_STR_PERMANENT) && startup_done) {
```

Comment:
This means, that we won't allocate CE cache slots during requests if opcache active. Right?
This will affect eval()'ed code, opcache.blacklisted files and situation when opcache is full.

Reply:
Yes, this is the key tradeoff with this patch. There are cases where we can't allocate CE_CACHE (eval with opcache, permanent strings during request without opcache), so property types will be uncached (arg types still have separate cache slot).

I didn't have any simple ideas on how to address this. One thought I had is that we could change the map_ptr mechanism to not work using "ptr or offset", but to have two different map_ptr_base for permanent and per-request and always use offset. That is `*(map_ptr_base[ptr & 1] + ptr)` as the access method. Per-request slots would then be allocated in the map_ptr area rather than on CG arena. Though just that isn't enough as well, as with opcache there are no per-request interned strings at all.

Do you have any ideas?

Reply:
It's possible to grow map_ptr vector in both directions having the single biased base. e.g. positive offsets for permanent slots and negative for per-request slots. anyway, this is not for 8.1